arangod/Pregel/Algos/HITSKleinberg.cpp

```diff
@@ -0,0 +1,331 @@
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Simon Grätzer
+////////////////////////////////////////////////////////////////////////////////
+
+#include "HITSKleinberg.h"
+#include <cmath>
+#include "Pregel/Aggregator.h"
+#include "Pregel/Algorithm.h"
+#include "Pregel/GraphStore.h"
+#include "Pregel/MasterContext.h"
+#include "Pregel/VertexComputation.h"
+
+using namespace arangodb::pregel::algos;
+using namespace arangodb;
+using namespace arangodb::pregel;
+
+static std::string const authAggregator = "auth";
+static std::string const hubAggregator = "hub";
+static std::string const maxDiffAuthAggregator = "diffA";
+static std::string const maxDiffHubAggregator = "diffH";
+static std::string const isLastIterationAggregator = "stop";
+
+using VertexType = HITSKleinbergValue;
+
+static double const epsilon = 0.00001;
+
+enum class State {
+  sendInitialHubs,
+  updateAuth,
+  updateAuthNormalizeHub,
+  updateHubNormalizeAuth,
+  finallyNormalizeHubs,
+  finallyNormalizeAuths
+};
+
+struct HITSKleinbergWorkerContext : public WorkerContext {
+  HITSKleinbergWorkerContext(size_t maxGSS, size_t numIterations)
+      : numIterations(numIterations){};
+
+  double authDivisor = 0;
+  double hubDivisor = 0;
+  State state = State::sendInitialHubs;
+  size_t numIterations;
+  size_t currentIteration = 0;
+
+  void preGlobalSuperstep(uint64_t gss) override {
+    // todo: we only need this if the global super-step is odd
+    //  difficulty: the parent class WorkerContext doesn't know the current gss
+    auto const* authNorm = getAggregatedValue<double>(authAggregator);
+    auto const* hubNorm = getAggregatedValue<double>(hubAggregator);
+    authDivisor = std::sqrt(*authNorm);
+    hubDivisor = std::sqrt(*hubNorm);
+
+    double const authMaxDiff =
+        *getAggregatedValue<double>(maxDiffAuthAggregator);
+    double const hubMaxDiff = *getAggregatedValue<double>(maxDiffHubAggregator);
+    const double diff = std::max(authMaxDiff, hubMaxDiff);
+
+    if (diff < epsilon) {
+      if (state == State::updateAuthNormalizeHub) {
+        state = State::finallyNormalizeHubs;
+      } else if (state == State::updateHubNormalizeAuth) {
+        state = State::finallyNormalizeAuths;
+      }
+    }
+  }
+
+  void postGlobalSuperstep(uint64_t gss) override {
+    switch (state) {
+      case State::sendInitialHubs:
+        state = State::updateAuth;
+        break;
```

Comment:
```suggestion
      case State::sendInitialHubs: {
        state = State::updateAuth;
        } break;
```

Reply:
I usually suggest to use a scope in a switch statement

Reply:
I don't mind using a scope, just out of the interest: why?